eZ/Publish/API/Repository/PermissionCriterionResolver.php

```diff
@@ -0,0 +1,28 @@
+<?php
+
+/**
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+namespace eZ\Publish\API\Repository;
+
+/**
+ * This service provides methods for resolving criterion permissions.
+ */
+interface PermissionCriterionResolver
+{
+    /**
+     * Get criteria representation for a permission.
+     *
+     * Will return a criteria if current user has limited access to the given module/function,
+     * however if user has either full or no access then boolean is returned.
+     *
+     * @uses \eZ\Publish\API\Repository\PermissionResolver::hasAccess()
+     *
+     * @param string $module
+     * @param string $function
+     *
+     * @return bool|\eZ\Publish\API\Repository\Values\Content\Query\Criterion
```

Comment:
Maybe, but it's kind of needed here, there are three cases: full access, no access and limited access.

Alternative would be to rather place `addPermissionCriterion(&$criterion) : bool` in API instead, assuming LocationService can be adapted to use it.

It will swap the criterion if needed, and return false if user does not have access.
What it looses is ability to say user has full access, unless `$criterion` argument is allowed to be null or something hinting no change was made.

Example: Location service when it needs to check if you have permission to delete a subtree, where it does a negative match to see if anything in the tree is something you don't have access to.

Reply:
Yeah, I've seen how complex this is, so I'm not suggesting it as a change here :)

Still, in the future we should redesign this so this multi-type return is not needed (if it's needed then this implies bad design somewhere as *normal* languages won't allow it :grinning: ;) ).

I'm rather thinking of returning some specific Criterions instead of `true` and `false` and handle them properly.